lib/HLSL/DxilValidation.cpp

```diff
@@ -5372,6 +5380,211 @@ static void ValidateEntrySignatures(ValidationContext &ValCtx) {
+  }
+
+  void DiagnoseConflicts(Function *F, uint32_t conflictMask) {
+    // Emit a diagnostic indicating that either the entry function or a function
+    // called by the entry function contains a disallowed operation.
+    if (F == EntryFn)
+      ValCtx.EmitFnError(EntryFn, ValidationRule::SmIncompatibleOperation);
+    else
+      ValCtx.EmitFnError(EntryFn, ValidationRule::SmIncompatibleCallInEntry);
+
+    // Emit diagnostics for each conflict found in this function.
+    Diagnose(F, conflictMask, ConflictKind::Stage,
+             ValidationRule::SmIncompatibleStage,
+             {ShaderModel::GetKindName(props.shaderKind)});
+    Diagnose(F, conflictMask, ConflictKind::ShaderModel,
+             ValidationRule::SmIncompatibleShaderModel);
+    Diagnose(F, conflictMask, ConflictKind::DerivLaunch,
+             ValidationRule::SmIncompatibleDerivLaunch,
+             {GetLaunchTypeStr(props.Node.LaunchType)});
+    Diagnose(F, conflictMask, ConflictKind::DerivThreadGroupDim,
+             ValidationRule::SmIncompatibleThreadGroupDim,
+             {std::to_string(props.numThreads[0]),
+              std::to_string(props.numThreads[1]),
+              std::to_string(props.numThreads[2])});
+    Diagnose(F, conflictMask, ConflictKind::DerivInComputeShaderModel,
+             ValidationRule::SmIncompatibleDerivInComputeShaderModel);
+    Diagnose(F, conflictMask, ConflictKind::RequiresGroup,
+             ValidationRule::SmIncompatibleRequiresGroup);
+  }
+
+  // Visit function and all functions called by it.
+  // Emit diagnostics for incompatibilities found in a function when no
+  // called functions introduced the conflict.
+  // Return conflict mask for this function.
+  uint32_t Visit(Function *F, uint32_t &remainingMask,
+                 llvm::SmallPtrSet<Function *, 8> &visited, CallGraph &CG) {
+    // Recursive check looks for where a conflict is found and not present
+    // in functions called by the current function.
+    // - When a source is found, emit diagnostics and clear the conflict
+    // flags introduced by this function from the working mask so we don't
+    // report this conflict again.
+    // - When the remainingMask is 0, we are done.
+
+    if (remainingMask == 0)
+      return 0; // Nothing left to search for.
+    if (!visited.insert(F).second)
+      return 0; // Already visited.
+
+    const DxilModule::ShaderCompatInfo *compatInfo =
+        ValCtx.DxilMod.GetCompatInfoForFunction(F);
+    DXASSERT(compatInfo, "otherwise, compat info not computed in module");
+    if (!compatInfo)
+      return 0;
+    uint32_t maskForThisFunction = IdentifyConflict(*compatInfo);
+
+    uint32_t maskForCalls = 0;
+    if (CallGraphNode *CGNode = CG[F]) {
+      for (auto &Call : *CGNode) {
+        Function *called = Call.second->getFunction();
+        if (called->isDeclaration())
+          continue;
+        maskForCalls |= Visit(called, remainingMask, visited, CG);
+        if (remainingMask == 0)
```

Comment:
Nit: It seems more natural to me to return the value that continued recursion depends on and pass the other value back as a parameter. That way you wouldn't need to `|=` it and could just modify it by reference.

Reply:
So your suggestion is to add the `maskForCalls` as a reference param, which gets ORd in the Visit call, and `remainingMask` will be passed in by-value, then Visit will return the masked result after diagnosing a conflict?

That's one additional parameter and doesn't really seem to simplify things, since you would need to track two separate values for `maskForCalls` inside Visit: one for the parameter which you OR bits for this functions into, and a new one on the stack for the inner calls to Visit, which is initialized to 0 and gets OR'd into by those inner calls.  That spreads the logic for `maskForCalls` between an outer and inner Visit call, which seems more complicated than just doing it where we want to see the union of the conflict masks for the inner calls.

The current definition returns a simple value: the function's conflict mask.  It also updates remainingMask for the remaining diagnostics.

I'm not sure I want to refactor this code at this point.  If I do, I would need to revisit the decision of whether or not to diagnose the same conflict multiple times when it shows up from different called functions.  This would also influence the approach with `remainingMask` as well.